File: framework/languages/java/e2e/SpotService/Shared/src/main/java/systems/zlink/e2e/spotservice/shared/ActorAuthHandler.java
package systems.zlink.e2e.spotservice.shared;

import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkTypedSessionPacketHandler;

public final class ActorAuthHandler
    implements ZLinkTypedSessionPacketHandler<ZLinkSessionContext, Contracts.ActorAuthReq> {
    private final ZLinkActorManager actors;
    private final ScenarioState evidence;

    public ActorAuthHandler(
        ZLinkActorManager actors,
        ScenarioState evidence) {
        this.actors = actors;
        this.evidence = evidence;
    }

    @Override
    public Class<Contracts.ActorAuthReq> messageType() {
        return Contracts.ActorAuthReq.class;
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> handle(
        ZLinkSessionContext context,
        ZLinkSessionDispatchContext dispatch,
        Contracts.ActorAuthReq request) {
        return actors.find(request.actorId())
            .thenCompose(existing -> existing.<java.util.concurrent.CompletionStage<ActorRef>>map(
                java.util.concurrent.CompletableFuture::completedFuture)
                .orElseGet(() -> actors.create(request.actorId(), "scenario")
                    .request(request)
                    .submit()
                    .thenApply(ActorAuthHandler::createdActor)))
            .thenCompose(actor -> context.actors().bind(actor))
            .thenAccept(bound -> {
                evidence.record("ActorSessionBound", "session", request.actorId());
                context.client().reply(new Contracts.ActorAuthRes(
                    bound.actorId(), bound.ref().nodeRid().toString(), bound.ref().objectGeneration(),
                    context.actors().bound().size(),
                    request.profile().displayName(), request.profile().level(), request.profile().tags())).submit();
            });
    }

    private static ActorRef createdActor(ZLinkActorCreateResult result) {
        if (result instanceof ZLinkActorCreateResult.Existing existing) {
            return existing.actor();
        }
        if (result instanceof ZLinkActorCreateResult.Created created) {
            return created.actor();
        }
        throw new IllegalStateException("actor creation was rejected");
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Session/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/session/handlers/ActorAuthHandler.kt
package systems.zlink.e2e.kotlin.spotservice.session.handlers

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.ScenarioState
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import kotlinx.coroutines.future.await
import systems.zlink.framework.kotlin.ZLinkSuspendingTypedSessionPacketHandler
import systems.zlink.framework.kotlin.kotlin

class ActorAuthHandler(
    private val actors: ZLinkActorManager,
    private val evidence: ScenarioState
) : ZLinkSuspendingTypedSessionPacketHandler<ZLinkSessionContext, Contracts.ActorAuthReq> {
    override fun packetName(): String = "ActorAuthReq"

    override fun messageType(): Class<Contracts.ActorAuthReq> = Contracts.ActorAuthReq::class.java

    override suspend fun handle(
        context: ZLinkSessionContext,
        dispatch: ZLinkSessionDispatchContext,
        request: Contracts.ActorAuthReq
    ) {
        val actor = when (val result = actors.kotlin()
            .getOrCreate(request.actorId, "scenario")
            .request(request)
            .await()) {
            is ZLinkActorCreateResult.Existing -> result.actor
            is ZLinkActorCreateResult.Created -> result.actor
            is ZLinkActorCreateResult.Rejected -> error("actor creation was rejected")
        }
        val bound = context.actors().bind(actor)
            .await()
        evidence.record("ActorSessionBound", "session", request.actorId)
        context.client()
            .reply(
                Contracts.ActorAuthRes(
                    bound.actorId(),
                    bound.ref().nodeRid().toString(),
                    context.actors().bound().size,
                    request.profile.displayName,
                    request.profile.level,
                    request.profile.tags
                )
            )
            .submit()
    }
}
